README.md
## Angular Activity Monitor
This is a simple service that will emit a couple of events based on the users' DOM activity. It also allows you to "keep items alive" in the background so long as the user is considered "active".

#### Installation:
```bash
$ [npm|bower] install --save angular-activity-monitor
```

#### Usage:
```js
// with bower (or without packaging)
angular.module('myModule', ['ActivityMonitor']);

// with npm (via webpack or Browserify)
angular.module('myModule', [require('angular-activity-monitor')]);

MyController.$inject = ['ActivityMonitor'];
function MyController(ActivityMonitor) {
  ActivityMonitor.on('inactive', function() {
    alert("y0, you're inactive!");
  });
}
```

##### `ActivityMonitor.options` (configuration):
 * `enabled`: whether to regularly check for inactivity (default: `false`) [bool]
 * `keepAlive`: background execution frequency (default: `800`) [seconds]
 * `inactive`: how long until user is considered inactive (default: `900`) [seconds]
 * `warning`: when user is nearing inactive state (deducted from inactive) (default: `60`) [seconds]
 * `disableOnInactive`: Once user is inactive, all event listeners are detached and activity monitoring is discontinued (default: `true`) [bool]
 * `DOMevents`: array of events on the DOM that count as user activity (default: `['mousemove', 'mousedown', 'mouseup', 'keypress', 'wheel', 'touchstart', 'scroll']`)

##### `ActivityMonitor.user` (information about the user):
 * `action`: timestamp of the users' last action (default: `Date.now()`) [milliseconds]
 * `active`: is the user considered active? (default: `true`) [bool]
 * `warning`: is the user nearing inactivity? (default: `false`) [bool]

##### `ActivityMonitor` Methods:
 * `on(event, callback)` (alias `bind`): subsribe to a particular event
 * `off(event[, callback])` (alias `unbind`): unsubscribe to a particular event. If no `callback` or `namespace` provided, all subscribers for the given `event` will be cleared.
 * `activity()`: manually invoke user activity (this updates the `User` object above)

##### `ActivityMonitor` Events:
 * `keepAlive`: anything to execute (at the `Options.keepAlive` interval) so long as the user is active.
 * `warning`: when user is approaching inactive state
 * `inactive`: when user is officially considered inactive
 
#### How long until user is inactive?
This can be configured by setting the `ActivityMonitor.options.inactive` property to the desired timeout (in seconds).

#### When is the user considered active?
Everytime one of the follow DOM events occur, the `action` and `active` properties on the `User` object is updated accordingly.
```js
var DOMevents = ['mousemove', 'mousedown', 'keypress', 'wheel', 'touchstart', 'scroll'];
```

#### (Un)subscribing and Event namespacing
If you've ever used [`jQuery.unbind()`](http://api.jquery.com/unbind/), you're in luck. This subscription model works _almost_ exactly like that. Subscribing is pretty straight forward using `.on()` or `.bind()` as described above but, unsubscribing gets a little weird. You essentially have two options:
 - Pass the same callback argument to `.unbind()` or `.off()`
 - Subscribe and unsubscribe using event namespacing.

_same callback example_
```js
var foo = function() {
  alert("y0, you're inactive!");
};
ActivityMonitor.on('inactive', foo); /* subscribe */
ActivityMonitor.off('inactive', foo); /* unsubscribe */
```

_event namespace example_

Instead of maintaining references to callbacks in order to unbind them, we can namespace the events and use this capability to easily unbind our actions. Namespaces are defined by using a period (`.`) character when binding to an event:
```js
ActivityMonitor.on('inactive.myEvent', function foo() {
  alert("y0, you're inactive!");
});
ActivityMonitor.off('inactive.myEvent');
```

If there's something missing or some quirk you've found. _FIX OR UPDATE ME!!!_


CHANGELOG.md
# Change Log
All notable changes to this project will be documented here.

## 1.1.0 - 2016-08-23
### Features
- Added CHANGELOG.md
- Added `disableOnInactive` option (default: `true`). (#10)
  - This allows the service to resume _after_ a user has been considered inactive.
- Exposed `ActivityMonitor.enable()` and `ActivityMonitor.disable()` for manual control of service. (#10)

### Bug Fix
- `DOMevents` is no longer "pre-compiled", allowing consumers to pass their own events (via options)


test/activity-monitor.spec.js
/* jshint -W117, -W030 */
'use strict';

describe('Angular Activity Monitor', function() {

  beforeEach(module('ActivityMonitor'));

    var ActivityMonitor;
    var noop = function() {};
    var options;

    beforeEach(inject(function(_ActivityMonitor_) {
        ActivityMonitor = _ActivityMonitor_;

        ActivityMonitor.options.keepAlive = 0.3; /* ping invterval (seconds) */
        ActivityMonitor.options.inactive = 0.4;  /* user is considered inactive (seconds) */
        ActivityMonitor.options.warning = 0.1;   /* warn user about inactivity (deducted from inactive in seconds) */
        ActivityMonitor.options.monitor = 0.2;   /* how frequently to check if the user is inactive (seconds) */

        options = ActivityMonitor.options;
    }));

    describe('when invoked', function() {
        it('should have predefined options', function() {
            expect(options).to.be.an.object;
        });

        it('should be disabled', function() {
            expect(options.enabled).to.equal(false);
        });

        it('should allow users to override options', function() {
            expect(options.keepAlive).to.not.equal(800);
        });

        it('should contain an object about the users\' activity', function() {
            expect(ActivityMonitor.user).to.be.an.object;
        });

        it('should consider the user active by default', function() {
            expect(ActivityMonitor.user.active).to.equal(true);
        });

        it('should allow consumer to subsribe to events', function() {
            expect(ActivityMonitor.on).to.be.a.function;
            expect(ActivityMonitor.bind).to.be.a.function;
            ActivityMonitor.on('keepAlive', noop);
            ActivityMonitor.on(); /* no error when missing event or callback  */
        });

        it('should allow consumer to unsubsribe to events', function() {
            expect(ActivityMonitor.off).to.be.a.function;
            expect(ActivityMonitor.unbind).to.be.a.function;
            ActivityMonitor.off('keepAlive', noop);
            ActivityMonitor.off('keepAlive'); /* clears entire namespace */
        });

        it('should allow consumers to invoke activity', function(done) {
            expect(ActivityMonitor.activity).to.be.a.function;

            setTimeout(function() {
                ActivityMonitor.activity();
                expect(ActivityMonitor.user.action).to.equal(Date.now());
                done();
            }, 2);
        });
    });

    describe('when subscribing to any event', function() {
        it('should enable the service', function() {
            expect(options.enabled).to.equal(false);
            ActivityMonitor.on('keepAlive', noop);
            expect(options.enabled).to.equal(true);
        });

        it('should broadcast events', function(done) {
            var now = Date.now();
            var threshold = options.keepAlive * 1000;

            ActivityMonitor.on('keepAlive', function() {
                expect(Date.now() - now).to.be.at.least(threshold);
                done();
            });
        });
    });

    describe('when broadcasting warning events', function() {
        it('should note that the user is inactive', function(done) {
            var now = Date.now();
            var threshold = (options.inactive - options.warning) * 1000;

            ActivityMonitor.on('warning', function() {
                expect(Date.now() - now).to.be.at.least(threshold);
                expect(ActivityMonitor.user.warning).to.equal(true);

                setTimeout(function() {
                    ActivityMonitor.activity();
                    expect(ActivityMonitor.user.warning).to.equal(false);
                    done();
                }, 1);
            });
        });
    });

    describe('when broadcasting inactive events', function() {
        it('should note that the user is inactive', function(done) {
            var now = Date.now();
            var threshold = options.inactive * 1000;

            ActivityMonitor.on('inactive', function() {
                expect(Date.now() - now).to.be.at.least(threshold);
                expect(ActivityMonitor.user.active).to.equal(false);
                done();
            });
        });

        it('should not broadcast inactive events during activity', function(done) {
            var now = Date.now();
            var threshold = (options.inactive * 1000) * 2;

            setTimeout(function() {
                ActivityMonitor.activity();
            }, (options.inactive - 0.1) * 1000);

            ActivityMonitor.on('inactive', function() {
                expect(Date.now() - now).to.be.at.least(threshold);
                done();
            });
        });
    });

    describe('when not disableOnInactive', function() {
        beforeEach(function(){
            ActivityMonitor.options.disableOnInactive = false
        })

        it('should be enabled',function(done){
            ActivityMonitor.on('keepAlive', noop);
            expect(options.enabled).to.equal(true);

            setTimeout(function() {
                expect(options.enabled).to.equal(true);
                done();
            }, 410);
        })

        it('should reactivate',function(done){
            ActivityMonitor.on('keepAlive', noop);
            expect(options.enabled).to.equal(true);

            setTimeout(function() {
                expect(options.enabled).to.equal(true);

                expect(typeof ActivityMonitor.enable.timer.inactivity).to.equal('undefined');
                expect(typeof ActivityMonitor.enable.timer.keepAlive).to.equal('undefined');

                ActivityMonitor.activity();

                expect(typeof ActivityMonitor.enable.timer.inactivity).to.equal('number');
                expect(typeof ActivityMonitor.enable.timer.keepAlive).to.equal('number');

                done();
            }, 410);
        })
    });

});


test/activity-monitor.service.spec.js
/* jshint -W117, -W030 */
'use strict';

describe('Angular Activity Monitor', function() {

  beforeEach(module('ActivityMonitor'));

    var ActivityMonitor;
    var noop = function() {};
    var options;

    beforeEach(inject(function(_ActivityMonitor_) {
        ActivityMonitor = _ActivityMonitor_;

        ActivityMonitor.options.keepAlive = 0.3; /* ping invterval (seconds) */
        ActivityMonitor.options.inactive = 0.4;  /* user is considered inactive (seconds) */
        ActivityMonitor.options.warning = 0.1;   /* warn user about inactivity (deducted from inactive in seconds) */
        ActivityMonitor.options.monitor = 0.2;   /* how frequently to check if the user is inactive (seconds) */

        options = ActivityMonitor.options;
    }));

    describe('when invoked', function() {
        it('should have predefined options', function() {
            expect(options).to.be.an.object;
        });

        it('should be disabled', function() {
            expect(options.enabled).to.equal(false);
        });

        it('should allow users to override options', function() {
            expect(options.keepAlive).to.not.equal(800);
        });

        it('should contain an object about the users\' activity', function() {
            expect(ActivityMonitor.user).to.be.an.object;
        });

        it('should consider the user active by default', function() {
            expect(ActivityMonitor.user.active).to.equal(true);
        });

        it('should allow consumer to subsribe to events', function() {
            expect(ActivityMonitor.on).to.be.a.function;
            expect(ActivityMonitor.bind).to.be.a.function;
            ActivityMonitor.on('keepAlive', noop);
            ActivityMonitor.on(); /* no error when missing event or callback  */
        });

        it('should allow consumer to unsubsribe to events', function() {
            expect(ActivityMonitor.off).to.be.a.function;
            expect(ActivityMonitor.unbind).to.be.a.function;
            ActivityMonitor.off('keepAlive', noop);
            ActivityMonitor.off('keepAlive'); /* clears entire namespace */
        });

        it('should allow consumers to invoke activity', function(done) {
            expect(ActivityMonitor.activity).to.be.a.function;

            setTimeout(function() {
                ActivityMonitor.activity();
                expect(ActivityMonitor.user.action).to.equal(Date.now());
                done();
            }, 2);
        });
    });

    describe('when subscribing to any event', function() {
        it('should enable the service', function() {
            expect(options.enabled).to.equal(false);
            ActivityMonitor.on('keepAlive', noop);
            expect(options.enabled).to.equal(true);
        });

        it('should broadcast events', function(done) {
            var now = Date.now();
            var threshold = options.keepAlive * 1000;

            ActivityMonitor.on('keepAlive', function() {
                expect(Date.now() - now).to.be.at.least(threshold);
                done();
            });
        });
    });

    describe('when broadcasting warning events', function() {
        it('should note that the user is inactive', function(done) {
            var now = Date.now();
            var threshold = (options.inactive - options.warning) * 1000;

            ActivityMonitor.on('warning', function() {
                expect(Date.now() - now).to.be.at.least(threshold);
                expect(ActivityMonitor.user.warning).to.equal(true);

                setTimeout(function() {
                    ActivityMonitor.activity();
                    expect(ActivityMonitor.user.warning).to.equal(false);
                    done();
                }, 1);
            });
        });
    });

    describe('when broadcasting inactive events', function() {
        it('should note that the user is inactive', function(done) {
            var now = Date.now();
            var threshold = options.inactive * 1000;

            ActivityMonitor.on('inactive', function() {
                expect(Date.now() - now).to.be.at.least(threshold);
                expect(ActivityMonitor.user.active).to.equal(false);
                done();
            });
        });

        it('should not broadcast inactive events during activity', function(done) {
            var now = Date.now();
            var threshold = (options.inactive * 1000) * 2;

            setTimeout(function() {
                ActivityMonitor.activity();
            }, (options.inactive - 0.1) * 1000);

            ActivityMonitor.on('inactive', function() {
                expect(Date.now() - now).to.be.at.least(threshold);
                done();
            });
        });
    });

});


karma.conf.js
'use strict';

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',

    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['mocha', 'chai'],

    // list of files / patterns to load in the browser
    files: [
      'node_modules/angular/angular.min.js',
      'node_modules/angular-mocks/angular-mocks.js',
      'activity-monitor.js',
      'test/*.js'
    ],

    // list of files to exclude
    exclude: [],

    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {},

    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],

    // web server port
    port: 9876,

    // enable / disable colors in the output (reporters and logs)
    colors: true,

    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,

    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: false,

    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['PhantomJS'],

    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: true,

    // Concurrency level
    // how many browser should be started simultaneous
    concurrency: Infinity
  })
}


javascripts/app.js
angular
  .module('app', ['ActivityMonitor'])
  .controller('myController', myController);

myController.$inject = ['ActivityMonitor', '$interval'];
function myController(ActivityMonitor, $interval) {
  var vm = this;
  var timeout = 10;
  var interval;

  vm.countdown = timeout;
  vm.startDemo = startDemo;

  /* demo defaults */
  ActivityMonitor.options.inactive = timeout;
  ActivityMonitor.options.monitor = 0.1;
  ActivityMonitor.options.warning = 2;
  
  function startDemo() {
    vm.countdown = timeout;
    ActivityMonitor.off('inactive.demo');
    ActivityMonitor.off('activity.demo');

    interval = $interval(function() {
      vm.countdown--;

      if (!vm.countdown) {
        $interval.cancel(interval);
      }

    }, 1000);

    ActivityMonitor.on('inactive.demo', function() {
      alert('You\'re inactive!');
    });

    ActivityMonitor.on('activity.demo', function() {
      vm.countdown = timeout;
    });
  }
}

src/activity-montior.service.js
/*jshint -W116, -W030, latedef: false */
'use strict';

(function() {
    angular
        .module('ActivityMonitor')
        .service('ActivityMonitor', ActivityMonitor);

    var MILLISECOND = 1000;
    var EVENT_KEEPALIVE = 'keepAlive';
    var EVENT_INACTIVE = 'inactive';
    var EVENT_WARNING = 'warning';

    /* @ngInject */
    function ActivityMonitor($document) {
        var service = this;

        /* configuration */
        service.options = {
            enabled: false, /* is the ActivityMonitor enabled? */
            keepAlive: 800, /* keepAlive ping invterval (seconds) */
            inactive: 900,  /* how long until user is considered inactive? (seconds) */
            warning: 60,    /* when to warn user when nearing inactive state (deducted from inactive in seconds) */
            monitor: 3      /* how frequently to check if the user is inactive (seconds) */
        };

        /* user activity */
        service.user = {
            action: Date.now(), /* timestamp of the users' last action */
            active: true,       /* is the user considered active? */
            warning: false      /* is the user in warning state? */
        };

        service.activity = activity;                /* method consumers can use to supply activity */
        service.on = service.bind = subscribe;      /* expose method to subscribe to events */
        service.off = service.unbind = unsubscribe; /* expose method to unsubscribe from events */

        var events = {};
        events[EVENT_KEEPALIVE] = {};  /* functions to invoke along with ping (options.frequency) */
        events[EVENT_INACTIVE] = {};   /* functions to invoke when user goes inactive (options.threshold) */
        events[EVENT_WARNING] = {};    /* functions to invoke when warning user about inactivity (options.warning) */

        var timer = {
            inactivity: null,   /* setInterval handle to determine whether the user is inactive */
            keepAlive: null     /* setInterval handle for ping handler (options.frequency) */
        };

        /* list of DOM events to determine user's activity */
        var DOMevents = ['mousemove', 'mousedown', 'keypress', 'wheel', 'touchstart', 'scroll'].join(' ');

        return service;

        ///////////////

        function disable() {
            service.options.enabled = false;

            clearInterval(timer.inactivity);
            clearInterval(timer.keepAlive);

            $document.off(DOMevents, activity);
        }

        function enable() {
            $document.on(DOMevents, activity);
            service.options.enabled = true;
            service.user.warning = false;

            timer.keepAlive = setInterval(function() {
                publish(EVENT_KEEPALIVE);
            }, service.options.keepAlive * MILLISECOND);

            timer.inactivity = setInterval(function() {
                var now = Date.now();
                var warning = now - ((service.options.inactive - service.options.warning) * MILLISECOND);
                var inactive = now - (service.options.inactive * MILLISECOND);

                /* should we display warning */
                if (!service.user.warning && service.user.action <= warning) {
                    service.user.warning = true;
                    publish(EVENT_WARNING);
                }

                /* should user be considered inactive? */
                if (service.user.active && service.user.action <= inactive) {
                    service.user.active = false;
                    publish(EVENT_INACTIVE);
                    disable();
                }
            }, service.options.monitor * MILLISECOND);
        }

        /* invoked on every user action */
        function activity() {
            service.user.active = true;
            service.user.action = Date.now();

            if (service.user.warning) {
                service.user.warning = false;
                publish(EVENT_KEEPALIVE);
            }
        }

        function publish(event) {
            if (!service.options.enabled) return;
            var spaces = Object.keys(events[event]);
            if (!event || !spaces.length) return;
            spaces.forEach(function(space) {
                events[event][space] && events[event][space]();
            });
        }

        function subscribe(event, callback) {
            if (!event || typeof callback !== 'function') return;
            event = _namespace(event, callback);
            events[event.name][event.space] = callback;
            !service.options.enabled && enable();
        }

        function unsubscribe(event, callback) {
            event = _namespace(event, callback);

            if (!event.space) {
                events[event.name] = {};
                return;
            }

            events[event.name][event.space] = null;
        }

        /* method to return event namespace */
        function _namespace(event, callback) {
            event = event.split('.');

            if (!event[1] && typeof callback === 'function') {
                /* if no namespace, use callback and strip all linebreaks and spaces */
                event[1] = callback.toString().substr(0, 150).replace(/\r?\n|\r|\s+/gm, '');
            }

            return {
                name: event[0],
                space: event[1]
            };
        }
    }
})();


activity-monitor.js
/*jshint -W116, -W030, latedef: false */
'use strict';

(function (root, factory) {
    if (typeof module !== 'undefined' && module.exports) {
        // CommonJS
        if (typeof angular === 'undefined') {
            factory(require('angular'));
        } else {
            factory(angular);
        }
        module.exports = 'ActivityMonitor';
    } else if (typeof define === 'function' && define.amd) {
        // AMD
        define(['angular'], factory);
    } else {
        // Global variables
        factory(root.angular);
    }
}(this, function (angular) {
    var m = angular
            .module('ActivityMonitor', [])
            .service('ActivityMonitor', ActivityMonitor);

    var MILLISECOND = 1000;
    var EVENT_KEEPALIVE = 'keepAlive';
    var EVENT_INACTIVE = 'inactive';
    var EVENT_WARNING = 'warning';
    var EVENT_ACTIVITY = 'activity';

    ActivityMonitor.$inject = ['$document'];
    function ActivityMonitor($document) {
        var service = this;

        /* configuration */
        service.options = {
            enabled: false, /* is the ActivityMonitor enabled? */
            keepAlive: 800, /* keepAlive ping invterval (seconds) */
            inactive: 900,  /* how long until user is considered inactive? (seconds) */
            warning: 60,    /* when to warn user when nearing inactive state (deducted from inactive in seconds) */
            monitor: 3,     /* how frequently to check if the user is inactive (seconds) */
            disableOnInactive: true, /* by default, once user becomes inactive, all listeners are detached */
            DOMevents: ['mousemove', 'mousedown', 'mouseup', 'keypress', 'wheel', 'touchstart', 'scroll'] /* list of DOM events to determine user's activity */
        };

        /* user activity */
        service.user = {
            action: Date.now(), /* timestamp of the users' last action */
            active: true,       /* is the user considered active? */
            warning: false      /* is the user in warning state? */
        };

        service.activity = activity;                /* method consumers can use to supply activity */
        service.on = service.bind = subscribe;      /* expose method to subscribe to events */
        service.off = service.unbind = unsubscribe; /* expose method to unsubscribe from events */

        var events = {};
        events[EVENT_KEEPALIVE] = {};  /* functions to invoke along with ping (options.frequency) */
        events[EVENT_INACTIVE] = {};   /* functions to invoke when user goes inactive (options.threshold) */
        events[EVENT_WARNING] = {};    /* functions to invoke when warning user about inactivity (options.warning) */
        events[EVENT_ACTIVITY] = {};   /* functions to invoke any time a user makes a move */

        var timer = {
            inactivity: null,   /* setInterval handle to determine whether the user is inactive */
            keepAlive: null     /* setInterval handle for ping handler (options.frequency) */
        };

        enable.timer = timer;
        service.enable = enable;
        service.disable = disable;

        return service;

        ///////////////

        function disable() {
            service.options.enabled = false;

            disableIntervals()

            $document.off(service.options.DOMevents.join(' '), activity);
        }

        function disableIntervals(){
            clearInterval(timer.inactivity);
            clearInterval(timer.keepAlive);
            delete timer.inactivity;
            delete timer.keepAlive;
        }

        function enable() {
            $document.on(service.options.DOMevents.join(' '), activity);
            service.options.enabled = true;
            service.user.warning = false;

            enableIntervals();
        }

        function enableIntervals(){
            timer.keepAlive = setInterval(function () {
                publish(EVENT_KEEPALIVE);
            }, service.options.keepAlive * MILLISECOND);

            timer.inactivity = setInterval(function () {
                var now = Date.now();
                var warning = now - (service.options.inactive - service.options.warning) * MILLISECOND;
                var inactive = now - service.options.inactive * MILLISECOND;

                /* should we display warning */
                if (!service.user.warning && service.user.action <= warning) {
                    service.user.warning = true;
                    publish(EVENT_WARNING);
                }

                /* should user be considered inactive? */
                if (service.user.active && service.user.action <= inactive) {
                    service.user.active = false;
                    publish(EVENT_INACTIVE);

                    if(service.options.disableOnInactive){
                        disable();
                    }else{
                        disableIntervals();//user inactive is known, lets stop checking, for now
                        dynamicActivity = reactivate;//hot swap method that handles document event watching
                    }
                }
            }, service.options.monitor * MILLISECOND);
        }

        /* function that lives in memory with the intention of being swapped out */
        function dynamicActivity(){
            regularActivityMonitor();
        }

        /* after user inactive, this method is hot swapped as the dynamicActivity method in-which the next user activity reactivates monitors */
        function reactivate() {
            enableIntervals();
            dynamicActivity = regularActivityMonitor;
        }

        /* invoked on every user action */
        function activity(){
            dynamicActivity()
        }

        /* during a users active state the following method is called */
        function regularActivityMonitor() {
            service.user.active = true;
            service.user.action = Date.now();

            publish(EVENT_ACTIVITY);

            if (service.user.warning) {
                service.user.warning = false;
                publish(EVENT_KEEPALIVE);
            }
        }

        function publish(event) {
            if (!service.options.enabled) return;
            var spaces = Object.keys(events[event]);
            if (!event || !spaces.length) return;
            spaces.forEach(function (space) {
                events[event][space] && events[event][space]();
            });
        }

        function subscribe(event, callback) {
            if (!event || typeof callback !== 'function') return;
            event = _namespace(event, callback);
            events[event.name][event.space] = callback;
            !service.options.enabled && enable();
        }

        function unsubscribe(event, callback) {
            event = _namespace(event, callback);

            if (!event.space) {
                events[event.name] = {};
                return;
            }

            events[event.name][event.space] = null;
        }

        /* method to return event namespace */
        function _namespace(event, callback) {
            event = event.split('.');

            if (!event[1] && typeof callback === 'function') {
                /* if no namespace, use callback and strip all linebreaks and spaces */
                event[1] = callback.toString().substr(0, 150).replace(/\r?\n|\r|\s+/gm, '');
            }

            return {
                name: event[0],
                space: event[1]
            };
        }
    }

    return m;
}));